### ISSUE TITLE
raises regexp

### ISSUE THREAD

**Issue by pytestbot** [pytest-dev/pytest#372]
Originally reported by: **György Kiss (BitBucket: [kissgyorgy](http://bitbucket.org/kissgyorgy), GitHub: [kissgyorgy](http://github.com/kissgyorgy))**

----------------------------------------

I used assertRaisesRegexp in `unittest` and I miss it from `pytest`. As I don't want to use `unittest` anymore :D I wrote a context manager:

```
#!python

import pytest
import py

class raises_regexp(object):
    def __init__(self, exception, message):
        self.exception = exception
        self.message = message
        self.excinfo = None

    def __enter__(self):
        self.excinfo = object.__new__(py.code.ExceptionInfo)
        return self.excinfo

    def __exit__(self, exc_type, exc_val, exc_tb):
        __tracebackhide__ = True
        if exc_type is None:
            pytest.fail('DID NOT RAISE %s' % self.exception)

        self.excinfo.__init__((exc_type, exc_val, exc_tb))

        if not issubclass(self.excinfo.type, self.exception):
            pytest.fail('%s RAISED instead of %s' % (exc_type, self.exception))

        if self.message not in str(exc_val):
            pytest.fail('message "%s" not found in "%s"' % (self.message, str(exc_val)))

        return True

```

Usage:

```
#!python

def test_some_exception_thrown():
    with raises_regexp(ExpectedException, "message contained"):
       # code to test
```

I think others could benefit from something like this and would be nice to have it in pytest!

----------------------------------------
- Bitbucket: https://bitbucket.org/pytest-dev/pytest/issue/372

**Comment by RonnyPfannschmidt** [comment]
proposal;

```python
with pytest.raises.matching(ValueError, '.*Regex'):
  ...
```

**Comment by RonnyPfannschmidt** [comment]
the builtin solution should modify the context manager and use the regex condition based on the value of the regex attribute

**Comment by foxx** [comment]
I was almost tempted to use `pytest`, until I saw this issue :/ Any particular reason why this hasn't been added yet?

**Comment by The-Compiler** [comment]
Probably a mixture between "nobody has looked at it yet" and "there is a plugin for it" and "it's easy enough to do without pytest support":

```python
with pytest.raises(ValueError) as excinfo:
    func()

assert re.match(pattern, str(excinfo.value))
```

But since many people have said they'd appreciate such a feature, and nobody seems to have had time yet, I'm sure a contribution would we very welcome!

**Comment by nicoddemus** [comment]
I could tackle this, but would like to know what is the preferred interface:

1. From [pytest-raisesregexp](https://github.com/Walkman/pytest_raisesregexp):

  ```python
  with pytest.raises_regexp(ExpectedException, r".* 1560"):
        function_to_test(arg1)

  pytest.raises_regexp(ExpectedException, r".* 1560", function_to_test, arg1):
  ```

1. Proposed by @RonnyPfannschmidt:

  ```python
  with pytest.raises.matching(ValueError, 'regex'):
      function_to_test(arg1)
  
  pytest.raises.matching(ValueError, 'regex', function_to_test, arg1)
  ```

1. Any other suggestion? :smile:

What do you guys think?

**Comment by nicoddemus** [comment]
(Just posted [a tweet](https://twitter.com/nicoddemus/status/636742661664309248) about this).

**Comment by The-Compiler** [comment]
Another proposal from @hpk42 was to add a keyword argument to `pytest.raises`:

```python
with pytest.raises(ValueError, regex=r'...'):
      ...
```

That has the downside of being only usable in context manager form.

I'd still prefer this form, though, as I don't think the non-contextmanager form is used widely anyways. 

`pytest.raises.matching(...)` looks good, but I think it might be confusing for people when there's a function as an attribute of another function.

**Comment by sscherfke** [comment]
I would also prefer @hpk42 ’s proposal.

But in this case you'd have to check if the second argument is a non-keyword argument, which would mean that it is the function to be called.  In that case, the `regex` kwarg would be an argument for that function (and not for `raises()`). 

This seems a little dirty to me, but on the other hand, it would also make it possible to add additional keyword arguments like `func`, `args` and `kwargs` to do stuff like:

```python
# "regex" as argument for "myfunc":
pytest.raises(ValueError, myfunc, regex=r'spam')

# "regex" as argument for "raises":
pytest.raises(
    ValueError,
    regex=r'spam',
    func=myfunc,
    args=(23, 'foo'),
    kwargs={'spam': 42})
```

**Comment by nicoddemus** [comment]
> Another proposal from @hpk42 was to add a keyword argument to pytest.raises

Thanks for the reminder @The-Compiler! We can call this option **3**, if people want to reference the options by number:

```python
with pytest.raises(ValueError, regex=r'...'):
      function_to_test()
# no functional form supported
``` 

Initially I liked that proposal, but later I realized that having a special case where suddenly adding a keyword argument breaks the functional form would be surprising to people. Having this inconsistency doesn't seem too appealing to me, and we have other proposals that fit the requirements just as well.

Proposal **1** (`pytest.raises_regexp`) has the advantages of being similar to `unittest`'s `self.assertRaisesRegexp` and already having a plugin available which means there are test suites out there already using it. Going forward with this proposal would be just a matter of merging an existing plugin into the core (mostly).

Personally I think **2** (`pytest.raises.matching`) reads better, and while it might seem weird at first glance, I think people are used to it by now due to `mock.patch.object`.

So in the end, my vote would go to **1**, with **2** coming as a very close second.

**Comment by flub** [comment]
Since the oldest supported python version is now 2.6 I don't think supporting the functional-style is very important anymore.

Using `pytest.raises_regexp()` means projects using the existing plugin will probably need changes, so that's not a good plan either.

I'm not sure why `pytest.raises(ValueError, func, args, kwargs, regex=r'...')` would not work btw?  This would be fully backwards compatible, no?

I think out of the existing proposals I probably like the keyword option most, closely followed by `pytest.raises.matching()`.

And just to add to the confusion, I'll add another proposal:

    with pytest.raises(Exception) as excinfo:
        function_to_test()
    assert excinfo.match(r'...')

(the `assert` is optional there but I think it should be supported)  It essentially just saves you from doing an import but will also provide nicer messages.  I do like it because it matches how you catch exceptions normally, in an except statement you can also only ever specify the types you want to catch so it seems normal to keep that signature the same.  It is also a way more extensible approach then keeping adding to the signature, but that's a pretty weak argument.

**Comment by The-Compiler** [comment]
> Using `pytest.raises_regexp()` means projects using the existing plugin will probably need changes, so that's not a good plan either.

What changes? Wouldn't the plugin just be blacklisted and the functionality works just as before?

> I'm not sure why `pytest.raises(ValueError, func, args, kwargs, regex=r'...')` would not work btw? This would be fully backwards compatible, no?

The problem is that the signature is `pytest.raises(ExpectedException, func, *args, **kwargs)`, so `regex` would be a part of `**kwargs`. By the way, this function already has three different signatures:

* `pytest.raises(ExpectedException)` (as context manager)
* `pytest.raises(ExpectedException, func, *args, **kwargs)`
* `pytest.raises(ExpectedException, "func(*args, **kwargs)")`

If we now add a fourth one (but only in the context manager form!), that's going to be... a bad idea.

> And just to add to the confusion, I'll add another proposal:
>
> ```python
> with pytest.raises(Exception) as excinfo:
>     function_to_test()
> assert excinfo.match(r'...')
> ```
>
> (the assert is optional there but I think it should be supported)

I like that! :+1: to both.

**Comment by nicoddemus** [comment]
> ```python
> with pytest.raises(Exception) as excinfo:
>    function_to_test()
> assert excinfo.match(r'...')
> ```

Oh my! I love this. :heart_eyes: 

No need to introduce another function, backward compatible, easy to understand, supports both `with` and functional form, fits on how we do matching using `excinfo` anyway (`assert str(excinfo) == 'msg'`)... I see only benefits! To me this proposal wins hands down. :smile:

**Comment by ionelmc** [comment]
I like this one, although it's a bit crufty:
```
with pytest.raises(Exception) as excinfo:
   function_to_test()
assert excinfo.match(r'...')
```

This looks a bit better:
```
with pytest.raises_regex(Exception, r'...') as excinfo:
   function_to_test()
```

* `pytest.raises_regex` better than `pytest.raises.matches` - people won't expect that a callable would have that attribute. `pytest.raises_regex` is more obvious.
* similar to unittest - in theory people would just type `regex` and IDE autocompletes. Very easy and clear.
* already used? (the plugin)

**Comment by RonnyPfannschmidt** [comment]
lets add a match method to execinfo after pulling py.code into pytest, then we can pretty much pick the external facade afterwards

**Comment by HolgerPeters** [comment]
I prefer the `match` method of excinfo. Regular expressions for matching the error message tend to get kind of long and that makes them a bit awkward to use within `with .... :`. asserting them later on feels more natural.

**Comment by RonnyPfannschmidt** [comment]
with the match method inside of the excinfo

implementing all other convenience variants is trivial

**Comment by nicoddemus** [comment]
:+1: on implementing `match` in `ExceptionInfo`. :smile:

**Comment by nicoddemus** [comment]
Related to #103.

**Comment by RonnyPfannschmidt** [comment]
fixed in c578226d43e96d1cc86f14de0779acd96222f0e4
